In [1]:
import os
import duckdb
from dotenv import load_dotenv

load_dotenv()
con = duckdb.connect()

## 1. Install and attach Postgres to DuckDB

In [2]:
con.sql("INSTALL postgres")
con.sql("LOAD postgres")

con.sql("ATTACH '' AS pg (TYPE postgres, READ_ONLY)")

con.sql("SHOW ALL TABLES").show()
con.sql("SELECT * FROM pg.public.staging_price_paid_cleaned LIMIT 5").show()

┌──────────┬─────────┬──────────────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───────────┐
│ database │ schema  │             name             │                                                                                                                                    column_names                                                                                                                                    │                                                                      column_types                                                                      │ temporary │
│ varcha

## 2. Count rows per district

In [3]:
con.sql("""
    SELECT * FROM postgres_query('pg', '
        SELECT district_clean, count(*) AS n
        FROM staging_price_paid_cleaned
        GROUP BY 1
        ORDER BY 2 DESC
    ')
""").show(max_rows=40)

┌──────────────────────────┬────────┐
│      district_clean      │   n    │
│         varchar          │ int64  │
├──────────────────────────┼────────┤
│ BIRMINGHAM               │ 457948 │
│ LEEDS                    │ 419023 │
│ BRADFORD                 │ 274428 │
│ SHEFFIELD                │ 255816 │
│ MANCHESTER               │ 250323 │
│ CITY OF BRISTOL          │ 244848 │
│ LIVERPOOL                │ 224116 │
│ KIRKLEES                 │ 215117 │
│ EAST RIDING OF YORKSHIRE │ 214599 │
│ WANDSWORTH               │ 209848 │
│ CORNWALL                 │ 188582 │
│ CROYDON                  │ 186708 │
│ CARDIFF                  │ 186557 │
│ BROMLEY                  │ 186333 │
│ BRIGHTON AND HOVE        │ 172010 │
│ WAKEFIELD                │ 171716 │
│ BARNET                   │ 171189 │
│ WIRRAL                   │ 168736 │
│ WIGAN                    │ 167534 │
│ MILTON KEYNES            │ 166627 │
│      ·                   │     ·  │
│      ·                   │     ·  │
│      ·    

## 3. Save Bristol and Powys slices
Bristol is the dense urban slice; Powys is the rural one (many addresses with a house name and no street).

In [4]:
os.makedirs("data/raw", exist_ok=True)

slices = {
    "bristol": "CITY OF BRISTOL",
    "powys": "POWYS",
}

for name, district in slices.items():
    con.sql(f"""
        COPY (
            SELECT *
            FROM pg.public.staging_price_paid_cleaned
            WHERE district_clean = '{district}'
        ) TO 'data/raw/{name}.parquet' (FORMAT parquet)
    """)

## 4. Reload from file on fresh DuckDB instance

In [5]:
# Drop postgres attachment and read snapshot
con = duckdb.connect()
con.sql("CREATE VIEW slice AS SELECT * FROM 'data/raw/bristol.parquet'")

con.sql("DESCRIBE slice").show(max_rows=30)
con.sql("SELECT count(*) AS n FROM slice").show()

┌──────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│   column_name    │ column_type │  null   │   key   │ default │  extra  │
│     varchar      │   varchar   │ varchar │ varchar │ varchar │ varchar │
├──────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ transaction_id   │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ price            │ INTEGER     │ YES     │ NULL    │ NULL    │ NULL    │
│ date_of_transfer │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ property_type    │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ old_new          │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ duration         │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ ppd_category     │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ record_status    │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ postcode_clean   │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ paon_clean       │ VARC

## 5. NULL Rates per column

In [6]:
cols = [r[0] for r in con.sql("DESCRIBE slice").fetchall()]

null_rates = " UNION ALL ".join(
    f"""SELECT '{c}' AS col,
               count(*) FILTER (WHERE {c} IS NULL) AS nulls,
               round(100.0 * count(*) FILTER (WHERE {c} IS NULL) / count(*), 2) AS pct
        FROM slice"""
    for c in cols
)
con.sql(f"SELECT * FROM ({null_rates}) ORDER BY pct DESC").show(max_rows=30)

# Check for empty strings
text_cols = [c for c in cols if c.endswith("_clean") or c == "flat_identifier"]
empties = " UNION ALL ".join(
    f"SELECT '{c}' AS col, count(*) FILTER (WHERE {c} = '') AS empty_strings FROM slice"
    for c in text_cols
)
con.sql(empties).show(max_rows=20)

┌──────────────────┬────────┬────────┐
│       col        │ nulls  │  pct   │
│     varchar      │ int64  │ double │
├──────────────────┼────────┼────────┤
│ flat_identifier  │ 201307 │  82.22 │
│ saon_clean       │ 188696 │  77.07 │
│ locality_clean   │ 118688 │  48.47 │
│ postcode_clean   │    351 │   0.14 │
│ street_clean     │     62 │   0.03 │
│ date_of_transfer │      0 │    0.0 │
│ property_type    │      0 │    0.0 │
│ record_status    │      0 │    0.0 │
│ ppd_category     │      0 │    0.0 │
│ price            │      0 │    0.0 │
│ old_new          │      0 │    0.0 │
│ duration         │      0 │    0.0 │
│ town_city_clean  │      0 │    0.0 │
│ district_clean   │      0 │    0.0 │
│ paon_clean       │      0 │    0.0 │
│ county_clean     │      0 │    0.0 │
│ transaction_id   │      0 │    0.0 │
└──────────────────┴────────┴────────┘
  17 rows                  3 columns

┌─────────────────┬───────────────┐
│       col       │ empty_strings │
│     varchar     │     int64   

## 6. How much does NULL SAON matter?
Flats are the problem (27.9% missing SAON)

In [7]:
con.sql("""
    SELECT property_type,
           count(*) AS n,
           round(100.0 * count(*) FILTER (WHERE saon_clean IS NULL) / count(*), 1) AS pct_null_saon,
           round(100.0 * count(*) FILTER (WHERE flat_identifier IS NOT NULL) / count(*), 1) AS pct_flat_id
    FROM slice
    GROUP BY 1 ORDER BY 2 DESC
""").show()

┌───────────────┬────────┬───────────────┬─────────────┐
│ property_type │   n    │ pct_null_saon │ pct_flat_id │
│    varchar    │ int64  │    double     │   double    │
├───────────────┼────────┼───────────────┼─────────────┤
│ T             │ 102793 │          98.9 │         0.5 │
│ F             │  73795 │          27.9 │        57.0 │
│ S             │  51556 │          98.8 │         0.5 │
│ D             │  11910 │          98.2 │         1.1 │
│ O             │   4794 │          78.9 │        12.4 │
└───────────────┴────────┴───────────────┴─────────────┘



## 7. Find flat-heavy buildings

In [8]:
con.sql("""
    SELECT postcode_clean, paon_clean, street_clean,
           count(*) AS n_sales,
           count(DISTINCT saon_clean) AS n_saon
    FROM slice
    WHERE saon_clean IS NOT NULL
    GROUP BY ALL
    ORDER BY n_sales DESC
    LIMIT 10
""").show()

┌────────────────┬────────────────────┬──────────────────┬─────────┬────────┐
│ postcode_clean │     paon_clean     │   street_clean   │ n_sales │ n_saon │
│    varchar     │      varchar       │     varchar      │  int64  │ int64  │
├────────────────┼────────────────────┼──────────────────┼─────────┼────────┤
│ BS3 3NG        │ AIRPOINT           │ SKYPARK ROAD     │     338 │    309 │
│ BS2 0PL        │ MIDLAND MEWS 24    │ WATERLOO ROAD    │     189 │     71 │
│ BS1 3LY        │ 5102 APARTMENTS    │ ST JAMES BARTON  │     173 │     58 │
│ BS1 5JQ        │ THE CRESCENT       │ HANNOVER QUAY    │     170 │     93 │
│ BS1 5JR        │ THE CRESCENT       │ HANNOVER QUAY    │     167 │     86 │
│ BS1 5JP        │ THE CRESCENT       │ HANNOVER QUAY    │     165 │     93 │
│ BS16 3RT       │ PURDY COURT        │ NEW STATION ROAD │     163 │     70 │
│ BS1 3DJ        │ HORIZON            │ BROAD WEIR       │     156 │     87 │
│ BS1 4AU        │ CENTRAL QUAY NORTH │ BROAD QUAY       │     1

## 8. Pick row from above table and look at its transactions

In [9]:
con.sql("""
    SELECT saon_clean, flat_identifier, property_type, date_of_transfer, price
    FROM slice
    WHERE postcode_clean = 'BS3 3NG' AND paon_clean = 'AIRPOINT'    -- change the postcode_clean and paon_clean fields
    ORDER BY saon_clean, date_of_transfer
""").show(max_rows=60)

┌────────────┬─────────────────┬───────────────┬──────────────────┬────────┐
│ saon_clean │ flat_identifier │ property_type │ date_of_transfer │ price  │
│  varchar   │     varchar     │    varchar    │       date       │ int32  │
├────────────┼─────────────────┼───────────────┼──────────────────┼────────┤
│ 102        │ 102             │ F             │ 2017-05-08       │ 154500 │
│ 103        │ 103             │ F             │ 2016-03-24       │ 141000 │
│ 104        │ 104             │ F             │ 2014-11-28       │ 125000 │
│ 104        │ 104             │ F             │ 2017-07-07       │ 175000 │
│ 104        │ 104             │ F             │ 2019-12-19       │ 177000 │
│ 105        │ 105             │ F             │ 2010-10-08       │ 120000 │
│ 105        │ 105             │ F             │ 2019-07-09       │ 188000 │
│ 106        │ 106             │ F             │ 2014-10-24       │ 136000 │
│ 106        │ 106             │ F             │ 2016-09-30       │ 167500 │

In [10]:
con.sql("""
    CREATE OR REPLACE VIEW linkage_input AS
    SELECT *, coalesce(flat_identifier, saon_clean, '<NONE>') AS unit_key
    FROM slice
    WHERE NOT (property_type IN ('F', 'O') AND saon_clean IS NULL)
""")

con.sql("""
    CREATE OR REPLACE VIEW unit_unknown AS
    SELECT * FROM slice
    WHERE property_type IN ('F', 'O') AND saon_clean IS NULL
""")

con.sql("""
    SELECT (SELECT count(*) FROM slice)         AS slice_rows,
           (SELECT count(*) FROM linkage_input) AS linkage_rows,
           (SELECT count(*) FROM unit_unknown)  AS unit_unknown_rows
""").show()

┌────────────┬──────────────┬───────────────────┐
│ slice_rows │ linkage_rows │ unit_unknown_rows │
│   int64    │    int64     │       int64       │
├────────────┼──────────────┼───────────────────┤
│     244848 │       220458 │             24390 │
└────────────┴──────────────┴───────────────────┘



## 9. Export linkage input and unit-unknown for each slice
Same split as above, applied to every slice so Bristol (urban) and Powys (rural) go through identical rules.

In [11]:
for name in ["bristol", "powys"]:
    con.sql(f"""
        COPY (
            SELECT *, coalesce(flat_identifier, saon_clean, '<NONE>') AS unit_key
            FROM 'data/raw/{name}.parquet'
            WHERE NOT (property_type IN ('F', 'O') AND saon_clean IS NULL)
        ) TO 'data/raw/{name}_linkage_input.parquet' (FORMAT parquet)
    """)
    con.sql(f"""
        COPY (
            SELECT * FROM 'data/raw/{name}.parquet'
            WHERE property_type IN ('F', 'O') AND saon_clean IS NULL
        ) TO 'data/raw/{name}_unit_unknown.parquet' (FORMAT parquet)
    """)

    con.sql(f"""
        SELECT '{name}' AS slice,
               (SELECT count(*) FROM 'data/raw/{name}.parquet')               AS slice_rows,
               (SELECT count(*) FROM 'data/raw/{name}_linkage_input.parquet') AS linkage_rows,
               (SELECT count(*) FROM 'data/raw/{name}_unit_unknown.parquet')  AS unit_unknown_rows
    """).show()

┌─────────┬────────────┬──────────────┬───────────────────┐
│  slice  │ slice_rows │ linkage_rows │ unit_unknown_rows │
│ varchar │   int64    │    int64     │       int64       │
├─────────┼────────────┼──────────────┼───────────────────┤
│ bristol │     244848 │       220458 │             24390 │
└─────────┴────────────┴──────────────┴───────────────────┘

┌─────────┬────────────┬──────────────┬───────────────────┐
│  slice  │ slice_rows │ linkage_rows │ unit_unknown_rows │
│ varchar │   int64    │    int64     │       int64       │
├─────────┼────────────┼──────────────┼───────────────────┤
│ powys   │      56690 │        53119 │              3571 │
└─────────┴────────────┴──────────────┴───────────────────┘



## Key Findings

### 1. The same flat appears under two SAON spellings
In Airpoint (BS3 3NG, a large apartment block), 338 sales carry 309 distinct SAON strings but only 251 distinct `flat_identifier` values. 58 units (23% of the building) appear as both a bare number (`106`) and a prefixed form (`FLAT 106`). Without `flat_identifier`, each of those flats would receive two property IDs and its resales would never link.

*Open question:* the format may depend on the year of sale (in the rows I looked at, 2007-2009 sales used `FLAT 7xx` and later ones bare numbers). If so, the split falls between a unit's first sale and its resales, which are exactly the pairs the index needs. To be tested by looking at the share of `FLAT`-prefixed SAONs by year.

### 2. A same-day batch of discounted sales, all category A
25 Airpoint sales on 2009-04-17. Every row I inspected is category A, and nearly all are new builds. Five units sold for exactly £86,600 and several for exactly £118,000, against roughly £145k-£230k for neighbouring units in mid-2008. Probably a developer clearing unsold stock at a discount, recorded as ordinary sales (note the time period).

Filtering to category A will not remove this pattern, so the index may need to handle first sales of new builds separately. That is an econometrics decision for later, so `old_new` is carried through.

### 3. Category B is small
B is 12,431 of 244,848 rows (5.1%). That is too few to affect linkage, so B stays in the linkage input (property identity does not depend on transaction type, and dropping B would break ownership chains). Filtering by `ppd_category` happens downstream, so the flip detector can run with and without B.

### 4. Flats and `O` properties with no SAON are excluded from linkage
- 27.9% of flats (20,606 rows) have no SAON. At the address level, 347 addresses have six or more such sales (about 8 each), which almost certainly means several different flats sharing one address string.
- With NULL treated as "missing", a no-unit row gives no evidence against any neighbour. Clustering is transitive, so it could bridge two different flats into one property, creating a false repeat sale. A false merge is worse than a missed link, because it shows up as an outlier in the index and as a false positive in the flip detector.
- Decision: rows with `property_type IN ('F', 'O')` and no SAON go to a separate view (`unit_unknown`, 24,390 rows, 10.0% of the slice) and get no property ID. `O` is treated like flats because "no sub-unit" cannot safely be asserted for commercial or mixed-use property.
- All remaining rows get `unit_key = coalesce(flat_identifier, saon_clean, '<NONE>')`. For houses, `<NONE>` is a real fact (no sub-unit), so two houses at the same address agree on it.
- **Limitation:** these rows will be absent from the repeat-sales index and the flip detector. The share will vary by area.